### PR TITLE
ENH: `stats.chi2_contingency`: add `method` parameter

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-7099

#### What does this implement/fix?
This adds a `method` parameter to `scipy.stats.chi2_contingency` to address the need for more accurate tests of $R \times C$ contingency tables. It also paves the way to add a `method` argument and $R \times C$ support to `fisher_exact`.

#### Additional information
Most `MonteCarloMethod`s will be specified by the user providing an `rvs` callable, but this is rather complex to set up for a Monte Carlo version of `chi2_contingency`. Instead, I've added an additional `random_state` parameter, which is already used for `BootstrapMethod`s and `PermutationMethod`s. This is also natural to use in the other function that accepts a `MonteCarloMethod`, `pearsonr`.

The test changes are not actually all that extensive; I just added the separate functions to a class Consider reviewing with "Hide whitespace" marked.

 (*Update 8/13: Re: `fisher_exact` - it's been so long that I'm forgetting, but I think the difference is just that the statistic would be the PMF of the `random_table` distribution.*)

### CODE REVIEW

File: scipy/stats/contingency.py [hunk 1]
```diff
@@ -365,6 +396,80 @@ def chi2_contingency(observed, correction=True, lambda_=None):
...
+    rowsums, colsums = stats.contingency.margins(observed)
+    X = stats.random_table(rowsums.ravel(), colsums.ravel(), seed=random_state)
+    def rvs(size):
+        n_resamples = size[0]
+        return X.rvs(size=n_resamples).reshape(size)
+
+    expected = expected.ravel()
+
+    def statistic(table, axis):
+        return np.sum((table - expected.ravel())**2/expected, axis=axis)
```
Comment (mdhaber) [inline scipy/stats/contingency.py:467]:
```suggestion
        return np.sum((table - expected)**2/expected, axis=axis)
```

File: scipy/stats/_resampling.py [hunk 2]
```diff
@@ -2132,9 +2132,21 @@ class ResamplingMethod:
         the statistic. Batch sizes >>1 tend to be faster when the statistic
         is vectorized, but memory usage scales linearly with the batch size.
         Default is ``None``, which processes all resamples in a single batch.
+    random_state : {None, int, `numpy.random.Generator`,
+                    `numpy.random.RandomState`}, optional
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:2136]: @melissawm I think I saw you fixing one of these at the summit - how do you make a line break not change the formatting?

Or should I just include this all on one line and add the `noqa` at the end of the docstring?

Comment (melissawm) [reply scipy/stats/_resampling.py:2136]: Here's how: 

```suggestion
    random_state : {None, int, `numpy.random.Generator`, \
    `numpy.random.RandomState`}, optional
```

This works, but breaks the line for the rendered output as well:
Captura de imagem_20240619_191040

Comment (mdhaber) [reply scipy/stats/_resampling.py:2136]: Hmm. Looks like that just makes it part of the text description. ("optional" is regular font.)

Comment (melissawm) [reply scipy/stats/_resampling.py:2136]: I was pretty sure I had this figured out but now can't find the commit. Let me look again

Comment (lucascolley) [reply scipy/stats/_resampling.py:2136]: x-ref https://github.com/numpy/numpydoc/issues/87#issuecomment-1832867402

Comment (melissawm) [reply scipy/stats/_resampling.py:2136]: Just found the same ☝🏻  thanks @lucascolley I think that's exactly what we're hitting 😢

Comment (tupui) [reply scipy/stats/_resampling.py:2136]: I would consider naming this rng as it's going to be the new name for it as per the SPEC7. That would mean that you would directly only support the new way, but IMO a better user story as people will not need to update their code vs now they might use update it once for this and then again to have the new behavior.

https://github.com/scientific-python/specs/issues/180

**Review by tupui — COMMENTED** [review]
A comment on the RNG (GutHub mobile now forces to write something to make a comment...)

File: scipy/stats/_resampling.py (hunk 2 above)
Comment (mdhaber) [inline scipy/stats/_resampling.py:2136]:
```suggestion
    random_state : {None, int, `numpy.random.Generator`, `numpy.random.RandomState`}, optional
```

File: scipy/stats/_resampling.py [hunk 3]
```diff
@@ -2132,9 +2132,21 @@ class ResamplingMethod:
...
+
+        Pseudorandom number generator state used to generate resamples.
+
+        If `random_state` is already a ``Generator`` or ``RandomState``
+        instance, then that instance is used.
+        If `random_state` is an int, a new ``RandomState`` instance is used,
+        seeded with `random_state`.
+        If `random_state` is ``None`` (default), the
+        `numpy.random.RandomState` singleton is used.
     """
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:2146]:
```suggestion
    """  # noqa: E501
```

File: scipy/stats/contingency.py [hunk 4]
```diff
@@ -241,6 +254,26 @@ def chi2_contingency(observed, correction=True, lambda_=None):
...
+    Because `scipy.stats.contingency.chi2_contingency` performs a two-sided
+    test, the alternative hypothesis does not indicate the direction of the
+    effect. We can use `scipy.stats.contingency.odds_ratio` to support the
+    conclusion that aspirin *reduces* the risk of ischemic stroke.
+
+    Below are further examples showing how larger contingency tables can be
+    tested.
+
+    A two-way example (2 x 3):
+
```
Comment (mdhaber) [inline scipy/stats/contingency.py:276]: Oops; I think this is just a merge conflict.
```suggestion
```

File: scipy/stats/contingency.py [hunk 5]
```diff
@@ -279,8 +312,22 @@ def chi2_contingency(observed, correction=True, lambda_=None):
...
+    ...                   [17, 16]])
+    >>> res = chi2_contingency(obs, correction=False)
+    >>> ref = chi2_contingency(obs, correction=False, method=PermutationMethod())
+    >>> res.pvalue, ref.pvalue
+    (0.0614122539870913, 0.1074)  # may vary
+
     For a more detailed example, see :ref:`hypothesis_chi2_contingency`.
-    """
+
+"""
```
Comment (mdhaber) [inline scipy/stats/contingency.py:330]:
```suggestion
    """
```

File: scipy/stats/contingency.py [hunk 6]
```diff
@@ -320,6 +350,84 @@ def chi2_contingency(observed, correction=True, lambda_=None):
...
+
+
+def _untabulate(table):
+    # converts a contingency table to paired samples indicating the
+    # correspondence between row and column indices
+    r, c = table.shape
+    x, y = [], []
+    for i in range(r):
+        for j in range(c):
+            x += ([i] * table[i, j])
```
Comment (j-bowhay) [inline scipy/stats/contingency.py:386]: Is there a reason for `+=`? `.append` is more idiomatic and faster

**Review by j-bowhay — COMMENTED** [review]
Two comments and a merge conflict but otherwise this seems in good shape

File: scipy/stats/contingency.py [hunk 7]
```diff
@@ -141,7 +143,7 @@ def expected_freq(observed):
 )
 
 
-def chi2_contingency(observed, correction=True, lambda_=None):
+def chi2_contingency(observed, correction=True, lambda_=None, method=None):
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/contingency.py:146]: Definitely. That was an oversight.

File: scipy/stats/_resampling.py [hunk 8]
```diff
@@ -2156,13 +2157,30 @@ class MonteCarloMethod(ResamplingMethod):
...
+
+    def __init__(self, n_resamples=9999, batch=None, rvs=None, rng=None):
+        if (rvs is not None) and (rng is not None):
+            message = 'Use of `rvs` and `rng` are mutually exclusive.'
+            raise ValueError(message)
+
+        self.n_resamples = n_resamples
+        self.batch = batch
+        self.rvs = rvs
+        self.rng = rng
```
Comment (mdhaber) [inline scipy/stats/_resampling.py:2178]: Before the latest commit, this normalized `rng` with `np.random.default_rng` right here. However, the other `ResamplingMethod`s don't do that anymore, so I removed that and normalize it right before it is used.

File: scipy/stats/_stats_py.py [hunk 9]
```diff
@@ -4595,12 +4595,15 @@ def statistic(x, y, axis):
...
             return statistic
 
-        if method.rvs is None:
-            rng = np.random.default_rng()
-            method.rvs = rng.normal, rng.normal
+        # `monte_carlo_test` accepts an `rvs` tuple of callables, not an `rng`
+        # If the user specified an `rng`, replace it with the appropriate callables
+        method = method._asdict()
+        if (rng := method.pop('rng', None)) is not None:  # goo-goo g'joob
+            rng = np.random.default_rng(rng)
```
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4602]: Now normalize right before use.

File: scipy/stats/contingency.py [hunk 10]
```diff
@@ -320,6 +350,84 @@ def chi2_contingency(observed, correction=True, lambda_=None):
...
+def _untabulate(table):
+    # converts a contingency table to paired samples indicating the
+    # correspondence between row and column indices
+    r, c = table.shape
+    x, y = [], []
+    for i in range(r):
+        for j in range(c):
+            x.append([i] * table[i, j])
+            y.append([j] * table[i, j])
+    return np.concatenate(x), np.concatenate(y)
```
Comment (mdhaber) [inline scipy/stats/contingency.py:388]: Changed this as suggested.

File: scipy/stats/contingency.py [hunk 11]
```diff
@@ -320,6 +350,84 @@ def chi2_contingency(observed, correction=True, lambda_=None):
...
+def _chi2_monte_carlo_method(observed, expected, method):
+    method = method._asdict()
+
+    if method.pop('rvs', None) is not None:
+        message = ('If the `method` argument of `chi2_contingency` is an '
+                   'instance of `MonteCarloMethod`, its `rvs` attribute '
+                   'must be unspecified. Use the `MonteCarloMethod` `rng` argument '
+                   'to control the random state.')
+        raise ValueError(message)
+    rng = np.random.default_rng(method.pop('rng', None))
```
Comment (mdhaber) [inline scipy/stats/contingency.py:413]: Again, randomize right before use.

File: scipy/stats/contingency.py [hunk 12]
```diff
@@ -29,6 +29,8 @@
 from ._crosstab import crosstab
 from ._odds_ratio import odds_ratio
 from scipy._lib._bunch import _make_tuple_bunch
+from scipy._lib._util import check_random_state
```
Comment (mdhaber) [inline scipy/stats/contingency.py:32]:
```suggestion
```
Maybe merge with `[lint only]` if CI comes back green otherwise.

**Comment by mdhaber** [comment]
Looks like that did the trick. After this, would you like to add $R \times C$ support for `scipy.stats.fisher_exact`?  It's basically the same as what I did to `chi2_contingency` here, but it uses `stats.random_table.pmf` as the statistic. I can write it instead, if you'd prefer.

**Review by j-bowhay — APPROVED** [review]
Thanks!

**Comment by j-bowhay** [comment]
> Looks like that did the trick. After this, would you like to add R × C support for `scipy.stats.fisher_exact`? It's basically the same as what I did to `chi2_contingency` here, but it uses `stats.random_table.pmf` as the statistic. I can write it instead, if you'd prefer.

I think I'd prefer to review for this

**Pull request merged**